File: app/src/main/cpp/runtime/RuntimeController.cpp

#include "RuntimeController.h"
#include <android/log.h>
#include <set>

#define LOG_TAG "RuntimeController"
#define LOGI(...) __android_log_print(ANDROID_LOG_INFO, LOG_TAG, __VA_ARGS__)
#define LOGW(...) __android_log_print(ANDROID_LOG_WARN, LOG_TAG, __VA_ARGS__)

namespace render {

RuntimeController::RuntimeController() 
    : currentState_(RuntimeState::IDLE) {}

RuntimeState RuntimeController::currentState() const {
    std::lock_guard<std::mutex> lock(stateLock_);
    return currentState_;
}

Result RuntimeController::prepare(const std::string& source) {
    std::lock_guard<std::mutex> lock(stateLock_);
    logStateTransition("prepare", currentState_, RuntimeState::PREPARED);
    
    if (currentState_ == RuntimeState::RELEASED) {
        LOGW("prepare called on released engine");
        return Result::Failure(ErrorCode::ALREADY_RELEASED, "Engine already released");
    }
    
    if (currentState_ != RuntimeState::IDLE) {
        LOGW("Invalid state for prepare: %d", static_cast<int>(currentState_));
        return Result::Failure(ErrorCode::INVALID_STATE, "Invalid state");
    }
    
    currentState_ = RuntimeState::PREPARED;
    return Result::Success();
}

Result RuntimeController::play() {
    std::lock_guard<std::mutex> lock(stateLock_);
    logStateTransition("play", currentState_, RuntimeState::PLAYING);
    
    if (currentState_ == RuntimeState::RELEASED) {
        LOGW("play called on released engine");
        return Result::Failure(ErrorCode::ALREADY_RELEASED);
    }
    
    const std::set<RuntimeState> validStates = {
        RuntimeState::PREPARED,
        RuntimeState::PAUSED,
        RuntimeState::SEEK_RECOVERING
    };
    
    if (validStates.find(currentState_) == validStates.end()) {
        LOGW("Invalid state for play: %d", static_cast<int>(currentState_));
        return Result::Failure(ErrorCode::INVALID_STATE);
    }
    
    currentState_ = RuntimeState::PLAYING;
    return Result::Success();
}

Result RuntimeController::pause() {
    std::lock_guard<std::mutex> lock(stateLock_);
    logStateTransition("pause", currentState_, RuntimeState::PAUSED);
    
    if (currentState_ == RuntimeState::RELEASED) {
        LOGW("pause called on released engine");
        return Result::Failure(ErrorCode::ALREADY_RELEASED);
    }
    
    if (currentState_ != RuntimeState::PLAYING) {
        LOGW("Invalid state for pause: %d", static_cast<int>(currentState_));
        return Result::Failure(ErrorCode::INVALID_STATE);
    }
    
    currentState_ = RuntimeState::PAUSED;
    return Result::Success();
}

Result RuntimeController::seek(int64_t targetUs) {
    std::lock_guard<std::mutex> lock(stateLock_);
    logStateTransition("seek", currentState_, RuntimeState::SEEK_REQUESTED);
    
    if (currentState_ == RuntimeState::RELEASED) {
        LOGW("seek called on released engine");
        return Result::Failure(ErrorCode::ALREADY_RELEASED);
    }
    
    if (currentState_ != RuntimeState::PLAYING && currentState_ != RuntimeState::PAUSED) {
        LOGW("Invalid state for seek: %d", static_cast<int>(currentState_));
        return Result::Failure(ErrorCode::INVALID_STATE);
    }
    
    currentState_ = RuntimeState::SEEK_REQUESTED;
    return Result::Success();
}

Result RuntimeController::stop() {
    std::lock_guard<std::mutex> lock(stateLock_);
    logStateTransition("stop", currentState_, RuntimeState::STOPPED);
    
    if (currentState_ == RuntimeState::RELEASED) {
        LOGW("stop called on released engine");
        return Result::Failure(ErrorCode::ALREADY_RELEASED);
    }
    
    if (currentState_ == RuntimeState::STOPPED) {
        return Result::Success();
    }
    
    currentState_ = RuntimeState::STOPPED;
    return Result::Success();
}

Result RuntimeController::release() {
    std::lock_guard<std::mutex> lock(stateLock_);
    logStateTransition("release", currentState_, RuntimeState::RELEASED);
    
    if (currentState_ == RuntimeState::RELEASED) {
        LOGW("release called on already released engine");
        return Result::Failure(ErrorCode::ALREADY_RELEASED);
    }
    
    if (currentState_ != RuntimeState::STOPPED) {
        LOGW("Invalid state for release: %d, must be STOPPED first", 
             static_cast<int>(currentState_));
        return Result::Failure(ErrorCode::INVALID_STATE);
    }
    
    currentState_ = RuntimeState::RELEASED;
    return Result::Success();
}

void RuntimeController::logStateTransition(const std::string& action,
                                          RuntimeState from,
                                          RuntimeState to) const {
    LOGI("[%s] State transition: %d -> %d", 
         action.c_str(), 
         static_cast<int>(from), 
         static_cast<int>(to));
}

} // namespace render

File: app/src/main/kotlin/com/example/render_native_demo/runtime/RuntimeController.kt
package com.example.render_native_demo.runtime

import android.util.Log
import java.util.concurrent.locks.ReentrantLock

class RuntimeController {
    companion object {
        private const val TAG = "RuntimeController"
    }

    private var state = RuntimeState.IDLE
    private val lock = ReentrantLock()

    fun prepare(source: String): Result<Unit> {
        Log.i(TAG, "[prepare] called with: $source")
        lock.lock()
        try {
            if (state == RuntimeState.IDLE) {
                state = RuntimeState.PREPARED
                Log.i(TAG, "[prepare] State transition: ${RuntimeState.IDLE} -> ${RuntimeState.PREPARED}")
                return Result.success()
            } else {
                Log.e(TAG, "[prepare] Invalid state: $state")
                return Result.failure(ErrorCode.INVALID_STATE)
            }
        } finally {
            lock.unlock()
        }
    }

    fun play(): Result<Unit> {
        Log.i(TAG, "[play] called")
        lock.lock()
        try {
            return when (state) {
                RuntimeState.PREPARED,
                RuntimeState.PAUSED,
                RuntimeState.SEEK_RECOVERING,
                RuntimeState.EOS_HOLD -> {
                    transitionTo("play", RuntimeState.PLAYING)
                    Result.success()
                }
                RuntimeState.RELEASED -> {
                    Log.e(TAG, "[play] Engine already released")
                    Result.failure(ErrorCode.ALREADY_RELEASED)
                }
                else -> {
                    Log.e(TAG, "[play] Invalid state: $state")
                    Result.failure(ErrorCode.INVALID_STATE)
                }
            }
        } finally {
            lock.unlock()
        }
    }

    fun pause(): Result<Unit> {
        Log.i(TAG, "[pause] called")
        lock.lock()
        try {
            if (state == RuntimeState.PLAYING) {
                state = RuntimeState.PAUSED
                Log.i(TAG, "[pause] State transition: ${RuntimeState.PLAYING} -> ${RuntimeState.PAUSED}")
                return Result.success()
            } else {
                Log.e(TAG, "[pause] Invalid state: $state")
                return Result.failure(ErrorCode.INVALID_STATE)
            }
        } finally {
            lock.unlock()
        }
    }

    fun seek(targetUs: Long): Result<Unit> {
        Log.i(TAG, "[seek] called with: $targetUs")
        lock.lock()
        try {
            return when (state) {
                RuntimeState.PLAYING,
                RuntimeState.PAUSED,
                RuntimeState.EOS_HOLD -> {
                    val resumeState = if (state == RuntimeState.PAUSED) RuntimeState.PAUSED else RuntimeState.PLAYING
                    transitionTo("seek", RuntimeState.SEEK_REQUESTED)
                    transitionTo("seek", RuntimeState.SEEK_FLUSHING)
                    transitionTo("seek", RuntimeState.SEEK_PRIMING)
                    transitionTo("seek", RuntimeState.SEEK_RECOVERING)
                    transitionTo("seek", resumeState)
                    Log.i(TAG, "[seek] Seek to: ${targetUs / 1000} ms")
                    Result.success()
                }
                RuntimeState.RELEASED -> {
                    Log.e(TAG, "[seek] Engine already released")
                    Result.failure(ErrorCode.ALREADY_RELEASED)
                }
                else -> {
                    Log.e(TAG, "[seek] Invalid state: $state")
                    Result.failure(ErrorCode.INVALID_STATE)
                }
            }
        } finally {
            lock.unlock()
        }
    }

    fun stop(): Result<Unit> {
        Log.i(TAG, "[stop] called")
        lock.lock()
        try {
            return when (state) {
                RuntimeState.RELEASED -> {
                    Log.e(TAG, "[stop] Engine already released")
                    Result.failure(ErrorCode.ALREADY_RELEASED)
                }
                RuntimeState.STOPPED -> Result.success()
                else -> {
                    transitionTo("stop", RuntimeState.STOPPED)
                    Result.success()
                }
            }
        } finally {
            lock.unlock()
        }
    }

    fun release(): Result<Unit> {
        Log.i(TAG, "[release] called")
        lock.lock()
        try {
            return when (state) {
                RuntimeState.RELEASED -> {
                    Log.e(TAG, "[release] Engine already released")
                    Result.failure(ErrorCode.ALREADY_RELEASED)
                }
                RuntimeState.STOPPED -> {
                    transitionTo("release", RuntimeState.RELEASED)
                    Result.success()
                }
                else -> {
                    Log.e(TAG, "[release] Invalid state: $state, stop must run first")
                    Result.failure(ErrorCode.INVALID_STATE)
                }
            }
        } finally {
            lock.unlock()
        }
    }

    fun getState(): RuntimeState {
        lock.lock()
        try {
            return state
        } finally {
            lock.unlock()
        }
    }

    fun markEndOfStream(): Result<Unit> {
        lock.lock()
        try {
            return if (state == RuntimeState.PLAYING) {
                transitionTo("eos", RuntimeState.EOS_HOLD)
                Result.success()
            } else {
                Result.failure(ErrorCode.INVALID_STATE)
            }
        } finally {
            lock.unlock()
        }
    }

    private fun transitionTo(action: String, nextState: RuntimeState) {
        val previous = state
        state = nextState
        Log.i(TAG, "[$action] State transition: $previous -> $nextState")
    }
}
